platform/security/session-management-impl/src/main/java/org/codice/ddf/security/session/management/impl/SessionManagementServiceImpl.java

```diff
@@ -39,69 +45,120 @@
+    } catch (SecurityServiceException e) {
+      LOGGER.error("Failed to renew", e);
+      return Long.toString(timeLeft);
     }
-    return timeLeft;
+
+    if (renewedPrincipals != null) {
+      tokenHolder.setPrincipals(renewedPrincipals);
+    }
+    timeLeft = getTimeLeft(renewedPrincipals, session);
+    return Long.toString(timeLeft);
   }
 
   @Override
   public URI getInvalidate(HttpServletRequest request) {
     String requestQueryString = request.getQueryString();
     return URI.create(
         SystemBaseUrl.EXTERNAL
-            .constructUrl("/logout?noPrompt=true")
-            .concat(requestQueryString != null ? "&" + requestQueryString : ""));
+            .constructUrl("/logout")
+            .concat(requestQueryString != null ? "?" + requestQueryString : ""));
   }
 
-  private long getTimeLeft(SecurityTokenHolder securityToken) {
-    SecurityAssertionImpl securityAssertion =
-        new SecurityAssertionImpl(securityToken.getSecurityToken());
-    long time = securityAssertion.getNotOnOrAfter().getTime();
-    return Math.max(time - clock.millis(), 0);
+  private long getTimeLeft(PrincipalCollection principals, HttpSession session) {
+    long timeLeft = session.getMaxInactiveInterval() * 1000L;
+
+    if (principals != null) {
+      long earliestAssertionExpiration =
+          principals
+              .byType(SecurityAssertion.class)
+              .stream()
+              .map(SecurityAssertion::getNotOnOrAfter)
+              .map(Date::getTime)
+              .min(Comparator.naturalOrder())
+              .orElse(Long.MAX_VALUE);
+      timeLeft = Math.min(earliestAssertionExpiration - clock.millis(), timeLeft);
+    }
+
+    return Math.max(timeLeft, 0);
   }
 
-  private void doRenew(SecurityToken securityToken, SecurityTokenHolder tokenHolder)
-      throws SecurityServiceException {
-    SAMLAuthenticationToken samlToken =
-        new SAMLAuthenticationToken(securityToken.getPrincipal(), securityToken);
-    Subject subject = securityManager.getSubject(samlToken);
-    for (Object principal : subject.getPrincipals().asList()) {
-      if (principal instanceof SecurityAssertion) {
-        tokenHolder.setSecurityToken(((SecurityAssertion) principal).getSecurityToken());
+  private PrincipalCollection renewSecurityAssertions(
+      SecurityTokenHolder tokenHolder, HttpServletRequest request) throws SecurityServiceException {
+    AuthenticationToken token;
+
+    Collection<SecurityAssertion> assertions =
+        tokenHolder.getPrincipals().byType(SecurityAssertion.class);
+    if (assertions.isEmpty()) {
+      return null;
+    } else if (assertions.size() == 1) {
+      /*
+       * Only guest and SAML assertions can be renewed. If there's exactly one assertion and it's
+       * one of these types, renew it. Otherwise, ignore it.
+       */
+      SecurityAssertion assertion = assertions.iterator().next();
+      if (assertion instanceof SecurityAssertionDefault) {
```

Comment:
❓ Is this logic correct? SecurityAssertionDefault isn't necessarily a guest assertion

Reply:
Yes. The weight of the default security assertion is 0, which is how we signify a guest assertion. https://github.com/codice/ddf/pull/5105/files/0f1fc6bb8e04f8f57f71e19bd2dbe65f78ca7494#diff-9cb1321581a6395f2bd4f41522cb8bb5R50